parsers/pcap.go

```diff
@@ -0,0 +1,525 @@
+package parsers
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"math"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/google/gopacket/pcapgo"
+	"github.com/google/gopacket/tcpassembly"
+	"github.com/google/gopacket/tcpassembly/tcpreader"
+)
+
+const (
+	conjoinPcapKeyHeader   = "Internal-ALP-Pcap-Conjoin-Key"
+	timestampPcapKeyHeader = "Internal-ALP-Pcap-Timestamp-Key"
+)
+
+type PcapParser struct {
+	queryString    bool
+	qsIgnoreValues bool
+
+	resCh chan *http.Response // conjoined *http.Response
+}
+
+func NewPcapParser(r io.Reader, rawServerIPs []string, serverPort uint16, query, qsIgnoreValues bool) (Parser, error) {
+	h, err := pcapgo.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+
+	serverIPs := make([]net.IP, len(rawServerIPs))
+	for i, rawServerIP := range rawServerIPs {
+		serverIP := net.ParseIP(rawServerIP)
+		if serverIP == nil {
+			return nil, fmt.Errorf("failed to parse ip: %s", rawServerIP)
+		}
+
+		serverIPs[i] = serverIP
+	}
+
+	reqCh := make(chan *http.Request)
+	resCh := make(chan *http.Response)
+	go func() {
+		ps := gopacket.NewPacketSource(h, h.LinkType())
+		sf := newPcapHttpStreamFactory(reqCh, resCh, serverIPs, serverPort)
+		sp := tcpassembly.NewStreamPool(sf)
+		asmblr := tcpassembly.NewAssembler(sp)
+		readAndAssembleAllPackets(ps, asmblr)
+		sf.gracefulShutdown()
+	}()
+
+	conjoinedResCh := make(chan *http.Response)
+	go conjoinRequestAndResponse(reqCh, resCh, conjoinedResCh)
+
+	return &PcapParser{
+		queryString:    query,
+		qsIgnoreValues: qsIgnoreValues,
+
+		resCh: conjoinedResCh,
+	}, nil
+}
+
+func (j *PcapParser) Parse() (*ParsedHTTPStat, error) {
+	res := <-j.resCh
+	if res == nil {
+		return nil, io.EOF
+	}
+	req := res.Request
+
+	reqTimestamp, err := unixNanoStrToTime(req.Header.Get(timestampPcapKeyHeader))
+	if err != nil {
+		return nil, err
+	}
+
+	resTimestamp, err := unixNanoStrToTime(res.Header.Get(timestampPcapKeyHeader))
+	if err != nil {
+		return nil, err
+	}
+	resTime := resTimestamp.Sub(reqTimestamp)
+
+	var uri string
+	if j.queryString {
+		if j.qsIgnoreValues {
+			values := req.URL.Query()
+			for q := range values {
+				values.Set(q, "xxx")
+			}
+			req.URL.RawQuery = values.Encode()
+		}
+		uri = req.URL.String()
```

Comment:
By the way, don't it have to sort the order of the query parameters?
Currently it is not sorted to match this implementation: https://github.com/tkuchiki/alp/blob/main/parsers/parser.go#L167

Reply:
Thank you for reporting bugs!
You're right.
I will fix it with another PR.

Reply:
fixed by 9070b3f